File: app/src/main/java/com/ericneidhardt/dynamicsoundboard/notifications/PendingSoundNotificationBuilder.java
package com.ericneidhardt.dynamicsoundboard.notifications;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.PendingIntent;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.res.Resources;
import android.graphics.Bitmap;
import android.graphics.Point;
import android.media.MediaMetadataRetriever;
import android.net.Uri;
import android.os.Build;
import com.ericneidhardt.dynamicsoundboard.BaseActivity;
import com.ericneidhardt.dynamicsoundboard.R;
import com.ericneidhardt.dynamicsoundboard.mediaplayer.EnhancedMediaPlayer;
import com.ericneidhardt.dynamicsoundboard.misc.IntentRequest;

/**
 * File created by eric.neidhardt on 04.12.2014.
 */
public class PendingSoundNotificationBuilder extends Notification.Builder
{
	private String playerId;
	public String getPlayerId()
	{
		return this.playerId;
	}

	private int notificationId;
	public int getNotificationId()
	{
		return this.notificationId;
	}

	public PendingSoundNotificationBuilder(Context context, EnhancedMediaPlayer player)
	{
		this(context, player, player.getMediaPlayerData().getPlayerId().hashCode());
	}

	public PendingSoundNotificationBuilder(Context context, EnhancedMediaPlayer player, int notificationId)
	{
		this(context, player, notificationId, player.getMediaPlayerData().getLabel(), null);
	}

	public PendingSoundNotificationBuilder(Context context, EnhancedMediaPlayer player, int notificationId, String title, String message)
	{
		super(context);
		this.playerId = player.getMediaPlayerData().getPlayerId();
		this.notificationId = notificationId;

		int currentApiVersion = android.os.Build.VERSION.SDK_INT;
		boolean isLollipopStyleAvailable = currentApiVersion >= Build.VERSION_CODES.LOLLIPOP;

		this.setActionStop(context, isLollipopStyleAvailable);
		if (player.isPlaying())
		{
			this.setOngoing(true);
			this.setActionPause(context, isLollipopStyleAvailable);
		}
		else
		{
			this.setOngoing(false);
			this.setActionPlay(context, isLollipopStyleAvailable);
		}
		this.setActionFadeOut(context, isLollipopStyleAvailable);

		this.setSmallIcon(R.drawable.ic_stat_pending_sounds);
		this.setDeleteIntent(this.getPendingIntent(context, Constants.ACTION_DISMISS));
		this.setContentIntent(this.getOpenActivityIntent(context));

		this.setContentTitle(title);
		this.setContentText(message);
		this.setScaledLargeIcon(this.getLargeIcon(context, player.getMediaPlayerData().getUri()));

		if (isLollipopStyleAvailable)
			this.addStyleLollipop();
	}

	private Bitmap getLargeIcon(Context context, String uri)
	{
		MediaMetadataRetriever mediaDataReceiver = new MediaMetadataRetriever();
		mediaDataReceiver.setDataSource(context, Uri.parse(uri));

		Resources resources = context.getResources();
		int requiredHeight = resources.getDimensionPixelSize(android.R.dimen.notification_large_icon_height);
		int requiredWidth = resources.getDimensionPixelSize(android.R.dimen.notification_large_icon_width);

		byte [] data = mediaDataReceiver.getEmbeddedPicture();
		if (data != null)
		{
			Point size = BitmapUtil.getBitmapSize(data);
			int sampleSize = BitmapUtil.getSampleFactor(size.x, size.y, requiredWidth, requiredHeight);
			return BitmapUtil.getBitmap(data, sampleSize);
		}
		else
		{
			Point size = BitmapUtil.getBitmapSize(resources, R.drawable.ic_notification_large);
			int sampleSize = BitmapUtil.getSampleFactor(size.x, size.y, requiredWidth, requiredHeight);
			return BitmapUtil.getBitmap(context, R.drawable.ic_notification_large, sampleSize);
		}
	}

	private void setScaledLargeIcon(Bitmap bitmap)
	{
		if (bitmap == null)
			return;
		this.setLargeIcon(bitmap);
	}

	@TargetApi(Build.VERSION_CODES.LOLLIPOP)
	private void addStyleLollipop()
	{
		Notification.MediaStyle style = new Notification.MediaStyle();
		style.setShowActionsInCompactView(1); // index of action play/pause
		this.setStyle(style);
		this.setCategory(Notification.CATEGORY_TRANSPORT);
	}

	private void setActionStop(Context context, boolean isLollipopStyleAvailable)
	{
		this.addAction(R.drawable.ic_notification_stop, isLollipopStyleAvailable ? context.getString(R.string.notification_stop_sound) : "",
				this.getPendingIntent(context, Constants.ACTION_STOP));
	}

	private void setActionPause(Context context, boolean isLollipopStyleAvailable)
	{
		this.addAction(R.drawable.ic_notification_pause, isLollipopStyleAvailable ? context.getString(R.string.notification_pause_sound) : "",
				this.getPendingIntent(context, Constants.ACTION_PAUSE));
	}

	private void setActionPlay(Context context, boolean isLollipopStyleAvailable)
	{
		this.addAction(R.drawable.ic_notification_play, isLollipopStyleAvailable ? context.getString(R.string.notification_play_sound) : "",
				this.getPendingIntent(context, Constants.ACTION_PLAY));
	}

	private void setActionFadeOut(Context context, boolean isLollipopStyleAvailable)
	{
		this.addAction(R.drawable.ic_notification_fade_out, isLollipopStyleAvailable ? context.getString(R.string.notification_fade_out_sound) : "",
				this.getPendingIntent(context, Constants.ACTION_FADE_OUT));
	}

	private PendingIntent getPendingIntent(Context context, String action)
	{
		Intent intent = new Intent(action);
		intent.putExtra(Constants.KEY_PLAYER_ID, this.playerId);
		intent.putExtra(Constants.KEY_NOTIFICATION_ID, this.notificationId);
		return PendingIntent.getBroadcast(context, this.notificationId, intent, 0);
	}

	private PendingIntent getOpenActivityIntent(Context context)
	{
		Intent intent = new Intent(context, BaseActivity.class);
		return PendingIntent.getActivity(context, IntentRequest.NOTIFICATION_OPEN_ACTIVITY, intent, 0);
	}

	public static IntentFilter getNotificationIntentFilter()
	{
		IntentFilter filter = new IntentFilter();
		filter.addAction(Constants.ACTION_DISMISS);
		filter.addAction(Constants.ACTION_PLAY);
		filter.addAction(Constants.ACTION_PAUSE);
		filter.addAction(Constants.ACTION_STOP);
		return filter;
	}
}

File: app/src/main/kotlin/org/neidhardt/dynamicsoundboard/notifications/PendingSoundNotificationBuilder.kt
package org.neidhardt.dynamicsoundboard.notifications

import android.app.Notification
import android.app.PendingIntent
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Bitmap
import android.media.MediaMetadataRetriever
import android.net.Uri
import android.support.v7.app.NotificationCompat
import org.neidhardt.dynamicsoundboard.R
import org.neidhardt.dynamicsoundboard.mediaplayer.MediaPlayerController
import org.neidhardt.ui_utils.utils.AndroidVersion
import org.neidhardt.dynamicsoundboard.misc.IntentRequest
import org.neidhardt.dynamicsoundboard.soundactivity.SoundActivity

/**
 * File created by eric.neidhardt on 04.12.2014.
 */
fun getNotificationIntentFilter(): IntentFilter
{
	val filter = IntentFilter()
	filter.addAction(ACTION_DISMISS)
	filter.addAction(ACTION_PLAY)
	filter.addAction(ACTION_STOP)
	filter.addAction(ACTION_FADE_OUT)
	return filter
}

data class PendingSoundNotification(val notificationId: Int, var playerId: String, var notification: Notification)
{
	fun isPlaylistNotification(): Boolean = this.notificationId == NOTIFICATION_ID_PLAYLIST
}

class PendingSoundNotificationBuilder
@JvmOverloads constructor
(
		context: Context,
		player: MediaPlayerController,
		val notificationId: Int = player.mediaPlayerData.playerId.hashCode(),
		title: String = player.mediaPlayerData.label,
		message: String? = null
) : NotificationCompat.Builder(context)
{
	val playerId: String = player.mediaPlayerData.playerId

	init
	{
		val isLollipopStyleAvailable = AndroidVersion.IS_LOLLIPOP_AVAILABLE
		this.setActionStop(context, isLollipopStyleAvailable)
		if (player.isPlayingSound)
		{
			this.setOngoing(true)
			this.setActionFadeOut(context, isLollipopStyleAvailable)
		}
		else
		{
			this.setOngoing(false)
			this.setActionPlay(context, isLollipopStyleAvailable)
		}

		this.setSmallIcon(R.drawable.ic_stat_pending_sounds)
		this.setDeleteIntent(this.getPendingIntent(context, ACTION_DISMISS))
		this.setContentIntent(this.getOpenActivityIntent(context))

		this.setContentTitle(title)
		this.setContentText(message)
		this.setScaledLargeIcon(this.getLargeIcon(context, player.mediaPlayerData.uri))

		this.addMediaStyle()
	}

	private fun getLargeIcon(context: Context, uri: String): Bitmap
	{
		val mediaDataReceiver = MediaMetadataRetriever()
		mediaDataReceiver.setDataSource(context, Uri.parse(uri))

		val resources = context.resources
		val requiredHeight = resources.getDimensionPixelSize(android.R.dimen.notification_large_icon_height)
		val requiredWidth = resources.getDimensionPixelSize(android.R.dimen.notification_large_icon_width)

		val data = mediaDataReceiver.embeddedPicture
		if (data != null)
		{
			val size = getBitmapSize(data)
			val sampleSize = getSampleFactor(size.x, size.y, requiredWidth, requiredHeight)
			return getBitmap(data, sampleSize)
		} else
		{
			val size = getBitmapSize(resources, R.drawable.ic_notification_large)
			val sampleSize = getSampleFactor(size.x, size.y, requiredWidth, requiredHeight)
			return getBitmap(context, R.drawable.ic_notification_large, sampleSize)
		}
	}

	private fun setScaledLargeIcon(bitmap: Bitmap?)
	{
		if (bitmap == null)
			return
		this.setLargeIcon(bitmap)
	}

	private fun addMediaStyle()
	{
		val style = NotificationCompat.MediaStyle()
		style.setShowActionsInCompactView(1) // index of action play/pause
		this.setStyle(style)
	}

	private fun setActionStop(context: Context, isLollipopStyleAvailable: Boolean)
	{
		this.addAction(this.buildAction(R.drawable.ic_notification_stop,
				if (isLollipopStyleAvailable)
					context.getString(R.string.notification_stop_sound)
				else
					""
				, this.getPendingIntent(context, ACTION_STOP))
		)
	}

	private fun setActionPlay(context: Context, isLollipopStyleAvailable: Boolean)
	{
		this.addAction(this.buildAction(R.drawable.ic_notification_play,
				if (isLollipopStyleAvailable)
					context.getString(R.string.notification_play_sound)
				else
					""
				, this.getPendingIntent(context, ACTION_PLAY))
		)
	}

	private fun setActionFadeOut(context: Context, isLollipopStyleAvailable: Boolean)
	{
		this.addAction(this.buildAction( R.drawable.ic_notification_pause,
				if (isLollipopStyleAvailable)
					context.getString(R.string.notification_pause_sound)
				else
					""
				, this.getPendingIntent(context, ACTION_FADE_OUT))
		)
	}

	private fun buildAction(iconId: Int, label: String, intent: PendingIntent) : android.support.v4.app.NotificationCompat.Action
	{
		return android.support.v4.app.NotificationCompat.Action.Builder(iconId, label, intent).build()
	}

	private fun getPendingIntent(context: Context, action: String): PendingIntent
	{
		val intent = Intent(action)
		intent.putExtra(KEY_PLAYER_ID, this.playerId)
		intent.putExtra(KEY_NOTIFICATION_ID, this.notificationId)
		return PendingIntent.getBroadcast(context, this.getRequestCode(), intent, 0)
	}

	private fun getOpenActivityIntent(context: Context): PendingIntent
	{
		val intent = Intent(context, SoundActivity::class.java)
		return PendingIntent.getActivity(context, IntentRequest.NOTIFICATION_OPEN_ACTIVITY, intent, 0)
	}

	private fun getRequestCode(): Int
	{
		if (this.notificationId != NOTIFICATION_ID_PLAYLIST)
			return this.notificationId
		return Integer.toString(this.notificationId + this.playerId.hashCode()).hashCode()
	}
}
